internal/graph/lookupsubjects.go

```diff
@@ -495,114 +905,389 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
+		}
+
+		lastSubjectID, _ := frc.headSectionValue()
+		if lastSubjectID == "" {
+			return "", spiceerrors.MustBugf("got invalid cursor")
+		}
+
+		endingSubjectIDs.Add(lastSubjectID)
+	}
+
+	sortedSubjectIDs := endingSubjectIDs.AsSlice()
+	sort.Strings(sortedSubjectIDs)
+
+	if len(sortedSubjectIDs) == 0 {
+		return "", nil
+	}
+
+	return sortedSubjectIDs[len(sortedSubjectIDs)-1], nil
 }
 
-func (lse *lookupSubjectsExclusion) CompletedChildOperations() error {
-	var foundSubjects datasets.SubjectSetByResourceID
-	metadata := emptyMetadata
+// createFilteredAndLimitedResponse creates a filtered and limited (as is necessary via the cursor and limits)
+// version of the subjects, returning a DispatchLookupSubjectsResponse ready for publishing with just that
+// subset of results.
+func createFilteredAndLimitedResponse(
+	ci cursorInformation,
+	subjects map[string]*v1.FoundSubjects,
+	metadata *v1.ResponseMeta,
+) (*v1.DispatchLookupSubjectsResponse, func(), error) {
+	if subjects == nil {
+		return nil, func() {}, spiceerrors.MustBugf("nil subjects given to createFilteredAndLimitedResponse")
+	}
 
-	for index := 0; index < len(lse.collectors); index++ {
-		collector := lse.collectors[index]
-		results := datasets.NewSubjectSetByResourceID()
-		for _, result := range collector.Results() {
-			metadata = combineResponseMetadata(metadata, result.Metadata)
-			if err := results.UnionWith(result.FoundSubjectsByResourceId); err != nil {
-				return fmt.Errorf("failed to UnionWith under lookupSubjectsExclusion: %w", err)
+	afterSubjectID, _ := ci.headSectionValue()
+
+	// Filter down the subjects found by the cursor (if applicable) and the apply a limit.
+	filteredSubjectIDs := mapz.NewSet[string]()
+	for _, foundSubjects := range subjects {
+		for _, foundSubject := range foundSubjects.FoundSubjects {
+			// NOTE: wildcard is always returned, because it is needed by all branches, at all times.
+			if foundSubject.SubjectId == tuple.PublicWildcard || (afterSubjectID == "" || foundSubject.SubjectId > afterSubjectID) {
+				filteredSubjectIDs.Add(foundSubject.SubjectId)
 			}
 		}
+	}
 
-		if index == 0 {
-			foundSubjects = results
-		} else {
-			foundSubjects.SubtractAll(results)
-			if foundSubjects.IsEmpty() {
-				return nil
-			}
+	sortedSubjectIDs := filteredSubjectIDs.AsSlice()
+	sort.Strings(sortedSubjectIDs)
+
+	subjectIDsToPublish := make([]string, 0, len(sortedSubjectIDs))
+	subjectIDsToPublishWithoutWildcard := make([]string, 0, len(sortedSubjectIDs))
+
+	done := func() {}
+	for _, subjectID := range sortedSubjectIDs {
+		// Wildcards are always published, regardless of the limit.
+		if subjectID == tuple.PublicWildcard {
+			subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+			continue
 		}
+
+		ok := ci.limits.prepareForPublishing()
+		if !ok {
+			break
+		}
+
+		subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+		subjectIDsToPublishWithoutWildcard = append(subjectIDsToPublishWithoutWildcard, subjectID)
```

Comment:
As far as I can tell this slice is only used to identify the latest subjectID so it can be used in the cursor. Why keeping track of all subjectIDs when we just could keep track of the last one?

Reply:
I was being lazy? Changed